packages/server/integrations/IntegrationServerManager.ts

```diff
@@ -0,0 +1,102 @@
+import {GraphQLResolveInfo} from 'graphql'
+import fetch from 'node-fetch'
+import {stringify} from 'querystring'
+import {IntegrationProvider, IntegrationToken} from '../types/IntegrationProviderAndTokenT'
+
+interface OAuth2ServerResponse {
+  access_token: string
+  created_at?: string
+  error?: any
+  refresh_token?: string
+  scope: string
+}
+
+export interface SetupOAuth2ProviderParams {
+  code: string
+  redirectUri: string
+}
+
+interface IntegrationTokenState
+  extends Pick<IntegrationToken, 'accessToken' | 'oauthRefreshToken' | 'oauthScopes'> {
+  createdAt?: IntegrationToken['createdAt'] | null
+}
+
+abstract class IntegrationServerManager {
+  provider: IntegrationProvider
+  token: IntegrationTokenState | null
+
+  constructor(provider: IntegrationProvider) {
+    this.provider = provider
+    this.token = null
+  }
+
+  /**
+   * Retrieve an access token from an oauth2 given a code and redirect_uri
+   * @param params
+   * @returns access token materials
+   */
+  async setupOauth2Provider(params: SetupOAuth2ProviderParams) {
+    const {code, redirectUri} = params
+    const queryParams = {
+      client_id: this.provider.oauthClientId,
+      client_secret: this.provider.oauthClientSecret,
+      code,
+      grant_type: 'authorization_code',
+      redirect_uri: redirectUri
+    }
+
+    const headers = {
+      Accept: 'application/json',
+      'Content-Type': 'application/json'
+    }
+
+    const uri = `${this.provider.serverBaseUri}/oauth/token?${stringify(queryParams)}`
```

Comment:
Does most of the services have OAuth token request URI like this?

Reply:
This will almost certainly need to be parameterized